### ISSUE TITLE
Expecting Object (while executing) (exception location: /work/ArangoDB/arangod/RestHandler/RestCursorHandler.cpp:320)

### ISSUE THREAD

**Issue by spirit-team** [arangodb/arangodb#13572]
## My Environment

* __ArangoDB Version__:        3.7.8
* __Storage Engine__:             RocksDB
* __Deployment Mode__:       Single Server
* __Deployment Strategy__:   Manual Start in Docker
* __Configuration__:               N/A
* __Infrastructure__:               own
* __Operating System__:        RHEL 7.9
* __Total RAM in your machine__:        16GB
* __Disks in use__:        SAN
* __Used Package__:              Docker - official Docker library

## Component, Query & Data

__Affected feature__:
AQL query using web interface & arangosh

__AQL query (if applicable)__:
```
LET Assignment = ( 
FOR st IN PM_special_teams 
	FILTER st.assignment_type == "vuln_title_lowpriority" 
	AND st.cve_criteria != NULL 
		RETURN {
		    "team_assignment": st.team_assignment,
		    "cve_criteria": st.cve_criteria,
		    "prime_contact": st.prime_contact,
		    "comment": st.comment
		}
) 
FOR st IN Assignment 
    FOR vuln IN PM_vulnerabilities 
        FILTER REGEX_TEST(vuln.title, st.cve_criteria, true)
            UPDATE vuln WITH { 
                "special_team_title": st.team_assignment, 
                "prime_contact": st.prime_contact, 
                "comment": st.comment 
            } IN PM_vulnerabilities

```
__AQL explain and/or profile (if applicable)__:
Profile
```
Query String (659 chars, cacheable: false):
 LET Assignment = ( 
 FOR st IN PM_special_teams 
 	FILTER st.assignment_type == "vuln_title_lowpriority" 
 	AND st.cve_criteria != NULL 
 		RETURN {
 		    "team_assignment": st.team_assignment,
 		    "cve_criteria": st.cve_criteria,
 		    "prime_contact": st.prime_contact,
 		    "comment": st.comment
 		}
 ) 
 FOR st IN Assignment 
     FOR vuln IN PM_vulnerabilities 
         FILTER REGEX_TEST(vuln.title, st.cve_criteria, true)
             UPDATE vuln WITH { 
                 "special_team_title": st.team_assignment, 
                 "prime_contact": st.prime_contact, 
                 "comment": st.comment 
             } IN PM_vulnerabilities

Execution plan:
 Id   NodeType                       Est.   Comment
  1   SingletonNode                     1   * ROOT
 15   IndexNode                      1021     - FOR #18 IN PM_special_teams   /* hash index scan, projections: `comment`, `prime_contact`, `cve_criteria`, `team_assignment` */    FILTER (#18.`cve_criteria` != null)   /* early pruning */   
 13   CalculationNode                1021       - LET #14 = { "special_team_title" : #18.`team_assignment`, "prime_contact" : #18.`prime_contact`, "comment" : #18.`comment` }   /* simple expression */
 10   EnumerateCollectionNode   183804504       - FOR vuln IN PM_vulnerabilities   /* full collection scan, projections: `_key`, `title` */
 11   CalculationNode           183804504         - LET #12 = REGEX_TEST(vuln.`title`, #18.`cve_criteria`, true)   /* simple expression */   /* collections used: vuln : PM_vulnerabilities */
 12   FilterNode                183804504         - FILTER #12
 14   UpdateNode                        0         - UPDATE vuln WITH #14 IN PM_vulnerabilities 

Indexes used:
 By   Name              Type      Collection           Unique   Sparse   Selectivity   Fields                                                     Ranges
 15   idx_12910479509   hash      PM_special_teams     false    false        98.29 %   [ `assignment_type`, `server_criteria`, `cve_criteria` ]   (#18.`assignment_type` == "vuln_title_lowpriority")
 14   primary           primary   PM_vulnerabilities   true     false       100.00 %   [ `_key` ]                                                 vuln

Functions used:
 Name         Deterministic   Cacheable   Uses V8
 REGEX_TEST   true            true        false  

Optimization rules applied:
 Id   RuleName
  1   inline-subqueries
  2   move-calculations-up
  3   move-filters-up
  4   remove-unnecessary-calculations
  5   move-calculations-up-2
  6   move-filters-up-2
  7   remove-data-modification-out-variables
  8   use-indexes
  9   remove-filter-covered-by-index
 10   patch-update-statements
 11   reduce-extraction-to-projection
 12   move-filters-into-enumerate

Write query options:
 Option                   Value
 waitForSync              false
 skipDocumentValidation   false
 keepNull                 true
 mergeObjects             true
 ignoreRevs               true
 isRestore                false
 ignoreErrors             false
 ignoreDocumentNotFound   false
 readCompleteInput        false
 consultAqlWriteFilter    false
 exclusive                false
 nullMeansRemove          false
```

__Dataset__:
PM_vulnerabilities
```
{
    "_key": "22534546472",
    "_id": "PM_vulnerabilities/22534546472",
    "_rev": "_b5Whp5y--g",
    "vulnerability_id": 170070,
    "nexpose_id": "oracle_linux-cve-2017-16644",
    "title": "Oracle Linux: (CVE-2017-16644) ELSA-2020-5837: Unbreakable Enterprise kernel security update",
    "date_published": "2017-11-07",
    "severity_score": 7,
    "severity": "Severe",
    "riskscore": 207.67282104492188,
    "cvss_score": 7.19999981,
    "exploits": 0,
    "exploitable": 0,
    "malware_kits": 0,
    "cvss_vector": "AV:L/AC:L/Au:N/C:C/I:C/A:C",
    "cvss_v3_vector": "CVSS:3.0/AV:P/AC:L/PR:L/UI:N/S:U/C:H/I:H/A:H",
    "cvss_v3_score": 6.6,
    "description": "The hdpvr_probe function in drivers/media/usb/hdpvr/hdpvr-core.c in the Linux kernel through 4.13.11 allows local users to cause a denial of service (improper error handling and system crash) or possibly have unspecified other impact via a crafted USB device.",
    "special_team_title": "UNIX",
    "comment": "2018-12-05 QA",
    "prime_contact": "John Doe"
  },
  {
    "_key": "22534546475",
    "_id": "PM_vulnerabilities/22534546475",
    "_rev": "_b5Whp5y--i",
    "vulnerability_id": 170073,
    "nexpose_id": "oracle_linux-cve-2020-12674",
    "title": "Oracle Linux: (CVE-2020-12674) (Multiple Advisories): dovecot security update",
    "date_published": "2020-08-12",
    "severity_score": 5,
    "severity": "Severe",
    "riskscore": 82.48636627197266,
    "cvss_score": 5,
    "exploits": 0,
    "exploitable": 0,
    "malware_kits": 0,
    "cvss_vector": "AV:N/AC:L/Au:N/C:N/I:N/A:P",
    "cvss_v3_vector": "CVSS:3.0/AV:N/AC:L/PR:N/UI:N/S:U/C:N/I:N/A:H",
    "cvss_v3_score": 7.5,
    "description": "In Dovecot before 2.3.11.3, sending a specially formatted RPA request will crash the auth service because a length of zero is mishandled.",
    "special_team_title": "UNIX",
    "comment": "2018-12-05 QA",
    "prime_contact": "John Doe"
  },
```

PM_special_teams
```
  {
    "_key": "22528758186",
    "_id": "PM_special_teams/22528758186",
    "_rev": "_b5Q7Ovu---",
    "assignment_type": "vuln_title_lowpriority",
    "cve_criteria": "database",
    "team_assignment": "Database",
    "prime_contact": "Jane Smith",
    "date": "7/17/2018",
    "QA": "Success",
    "server_criteria": ""
  },
  {
    "_key": "22528759579",
    "_id": "PM_special_teams/22528759579",
    "_rev": "_b5Q7OxW--Q",
    "assignment_type": "vuln_title_lowpriority",
    "cve_criteria": "Oracle Linux",
    "team_assignment": "UNIX",
    "prime_contact": "John Doe",
    "comment": "2018-12-05 QA",
    "date": "11/29/2018",
    "QA": "Success",
    "server_criteria": ""
  },
```

__Size of your Dataset on disk__:
[docker@server engine-rocksdb]$ pwd
/arangodb/arangodb3/engine-rocksdb
[docker@server engine-rocksdb]$ du -h
0       ./journals/archive
94M     ./journals
3.9G    .

__Replication Factor & Number of Shards (Cluster only)__:
N/A

## Steps to reproduce

1. Execute Query in web console or in arangosh
2. 
3. 

__Problem__:
2021-02-18T11:02:42Z [1] ERROR JavaScript exception in file '/usr/share/arangodb3/js/client/modules/@arangodb/arangosh.js' at 99,7:\
 ArangoError 4: Expecting Object (while executing) (exception location: /work/ArangoDB/arangod/RestHandler/RestCursorHandler.cpp:32\
0). Please report this error to arangodb.com\n!      throw error;\n!      ^\nstacktrace: ArangoError: Expecting Object (while execu\
ting) (exception location: /work/ArangoDB/arangod/RestHandler/RestCursorHandler.cpp:320). Please report this error to arangodb.com\\
n    at Object.exports.checkRequestResult (/usr/share/arangodb3/js/client/modules/@arangodb/arangosh.js:97:21)\n    at ArangoStatem\
ent.execute (/usr/share/arangodb3/js/client/modules/@arangodb/arango-statement.js:171:12)\n    at Proxy.ArangoDatabase._query (/usr\
/share/arangodb3/js/client/modules/@arangodb/arango-database.js:931:45)\n    at (command-line):1:4\n

__Expected result__:
Im not expecting this to run quickly but I would expecting it to be able to run and update 7878 documents in the PM_vulnerabilities collection.
It executes without issues on 3.6.11 on the MMFiles storage engine

If I Return a result instead of trying to Update, the query will run successfully.
Here's the Profile of the query when I do a Return instead of Update

```
Profiling information
A    0.001 ms initializingstartup time for query engine
B    0.090 ms parsingquery parsing
C    0.012 ms optimizing astabstract syntax tree optimizations
D    0.009 ms loading collectionsloading collections
E    0.035 ms instantiating planinstanciation of initial execution plan
F    0.205 ms optimizing planexecution plan optimization and permutation
G    2.270 s executingquery execution
H    0.016 ms finalizingquery finalization
```

**Comment by maxkernbach** [comment]
Hi @spirit-team,

I had no success reproducing the error with a 3.7.8 single server. It seems that the given two documents for both collections are not sufficient. Does the error show on your side when you run the query on the shared 4 documents?

If not, would it be possible to a create a dump (using [arangodump](https://www.arangodb.com/docs/stable/programs-arangodump.html)) which includes collections `PM_special_teams` and `PM_vulnerabilities`?.

You can send us a message to hackers@arangodb.com (this ML is not public) and attach the dump in that email. This way we can try to reproduce and find the root cause. Please reference the number of this issue in your email.

**Comment by spirit-team** [comment]
Hi @maxkernbach, thanks for the follow-up. I will try to run the query on version 3.7.10 just in case and I will see what I can do about those dumps, There is sensitive data in there so I cant just dump the collections as-is.

**Comment by jsteemann** [comment]
Not sure if related or not, but I just created a PR that fixes velocypack (vpack) errors that could be returned by the cursor API in case of an internal exception. In this case, the velocypack error could be somewhat unrelated and mask the original exception message. The original exception message was clobbered by the velocypack error message, which was triggered by trying to write the "error" and "code" attributes to the cursor result after the exception occurred. If the exception had happened during cursor result buildup, then the result could have been in a state that would not allow adding further attributes to it, and try to do so anyway could have led to follow-up errors.
Anyway, I can't say for sure if this is what has happened here, but it is not completely unlikely.

The PR in question for 3.7 is https://github.com/arangodb/arangodb/pull/13871

**Comment by Simran-B** [comment]
@spirit-team v3.7.11 is available. Can you please try this version and see if it resolves the issue or at least provides a better error message now?

**Comment by spirit-team** [comment]
@Simran-B 
Yep, already downloaded the new docker image yesterday, I'll test as as soon time allows and give an update on the error message or let you know if somehow it suddenly starts working

**Comment by spirit-team** [comment]
@Simran-B unfortunately the error is very similar with 3.7.11. This is what I am getting now.

`2021-05-06T18:57:45Z [155] ERROR [979b9] JavaScript exception in file '/usr/share/arangodb3/js/client/modules/@arangodb/arangosh.js' at 99,7: ArangoError 4: Expecting Object (while executing) (exception location: /work/ArangoDB/arangod/RestHandler/RestCursorHandler.cpp:333). Please report this error to arangodb.com\n!      throw error;\n!      ^\nstacktrace: ArangoError: Expecting Object (while executing) (exception location: /work/ArangoDB/arangod/RestHandler/RestCursorHandler.cpp:333). Please report this error to arangodb.com\n    at Object.exports.checkRequestResult (/usr/share/arangodb3/js/client/modules/@arangodb/arangosh.js:97:21)\n    at ArangoStatement.execute (/usr/share/arangodb3/js/client/modules/@arangodb/arango-statement.js:172:12)\n    at Proxy.ArangoDatabase._query (/usr/share/arangodb3/js/client/modules/@arangodb/arango-database.js:957:45)\n    at (command-line):1:4\n`

Any ideas on what else we can try here?

**Comment by jsteemann** [comment]
@spirit-team : ah, that's unfortunate.
If there is no way to share dataset in question, one way to make progress is to determine if the problem is cause by the selection or the update part of the query.

If you could run the query
```
LET Assignment = ( 
FOR st IN PM_special_teams 
	FILTER st.assignment_type == "vuln_title_lowpriority" 
	AND st.cve_criteria != NULL 
		RETURN {
		    "team_assignment": st.team_assignment,
		    "cve_criteria": st.cve_criteria,
		    "prime_contact": st.prime_contact,
		    "comment": st.comment
		}
) 
FOR st IN Assignment 
    FOR vuln IN PM_vulnerabilities 
        FILTER REGEX_TEST(vuln.title, st.cve_criteria, true)
        RETURN vuln
```
and it completes successfully, then the problem is likely in the update part of the query.

But if the selection-only query fails, the problem can hopefully be further narrowed down to a specific document, by adding LIMIT(s) to the query and then adjusting the LIMIT value(s) to find the offset for/after which the query begins to fail. Hopefully that allows pin-pointing the problematic document.

Thanks!

**Comment by spirit-team** [comment]
@jsteemann Yep, that query works. Returns 7878 documents in 4.411

```
A  0.005 ms  initializing  startup time for query engine
B  0.293 ms  parsing  query parsing
C  0.018 ms  optimizing ast  abstract syntax tree optimizations
D  0.014 ms  loading collections  loading collections
E  0.039 ms  instantiating plan  instanciation of initial execution plan
F  128.119 ms  optimizing plan  execution plan optimization and permutation
G  4.283 s  executing  query execution
H  0.019 ms  finalizing  query finalization
```

I tried sending the result of the LET Assignment = (...) in a temporary table with cve_criteria already in lowercase and with an index on that field.  And I also added another field in PM_vulnerabilities named lower_title, which is the content of title but already lowercased and I added an Index on that aswell. And then tried both of the following queries, but neither work in the end.

**Query 1**
```
FOR st IN PM_low_priority_assign
    FOR vuln IN PM_vulnerabilities
        FILTER vuln.lower_title LIKE st.cve_criteria
            UPDATE vuln WITH {
                "special_team_title": st.team_assignment,
                "prime_contact": st.prime_contact,
                "comment": st.comment
            } IN PM_vulnerabilities
```
**Query 2**            
```
FOR st IN PM_low_priority_assign
    FOR vuln IN PM_vulnerabilities
        FILTER REGEX_TEST(vuln.lower_title, st.cve_criteria, false)
            UPDATE vuln WITH {
                "special_team_title": st.team_assignment,
                "prime_contact": st.prime_contact,
                "comment": st.comment
            } IN PM_vulnerabilities
```

I also ran this query
```
LET Assignment = ( 
FOR st IN PM_special_teams 
	FILTER st.assignment_type == "vuln_title_lowpriority" 
	AND st.cve_criteria != NULL 
		RETURN {
		    "team_assignment": st.team_assignment,
		    "cve_criteria": st.cve_criteria,
		    "prime_contact": st.prime_contact,
		    "comment": st.comment
		}
) 
FOR st IN Assignment 
    FOR vuln IN PM_vulnerabilities 
        FILTER REGEX_TEST(vuln.title, st.cve_criteria, true)
        COLLECT item = st.cve_criteria WITH COUNT INTO Count
        RETURN {
            "cve_criteria": item,
            "count": Count
        }
```

This is the result of that query
```
cve_criteria | count
-- | --
database | 610
Domino | 251
IBM AIX | 840
jboss | 319
Jenkins | 461
Lotus Notes | 242
Oracle Linux | 4904
SharePoint | 246
tuxedo | 5
```

And finally I ran 2 seperate update query again but in the first I excluded "Oracle Linux" which as by far the biggest hit count, and in the second I only update the ones that match "Oracle Linux" and both upddate actually works. So individually they both work but together they dont =(
```
FOR st IN PM_low_priority_assign
    FOR vuln IN PM_vulnerabilities
        FILTER st.cve_criteria != "oracle linux" && REGEX_TEST(vuln.lower_title, st.cve_criteria, false)
            UPDATE vuln WITH {
                "special_team_title": st.team_assignment,
                "prime_contact": st.prime_contact,
                "comment": st.comment
            } IN PM_vulnerabilities
```
```
FOR st IN PM_low_priority_assign
    FOR vuln IN PM_vulnerabilities
        FILTER st.cve_criteria == "oracle linux" && REGEX_TEST(vuln.lower_title, st.cve_criteria, false)
            UPDATE vuln WITH {
                "special_team_title": st.team_assignment,
                "prime_contact": st.prime_contact,
                "comment": st.comment
            } IN PM_vulnerabilities
```
They respectively updated this:
6.156 s  3493 writes  0 writes ignored
1.217 s  4904 writes  0 writes ignored

I really dont get it. halp!

**Comment by jsteemann** [comment]
@spirit-team : ok, so the following query still fails with the original error?
```
FOR st IN PM_low_priority_assign
    FOR vuln IN PM_vulnerabilities
        FILTER vuln.lower_title LIKE st.cve_criteria
            UPDATE vuln WITH {
                "special_team_title": st.team_assignment,
                "prime_contact": st.prime_contact,
                "comment": st.comment
            } IN PM_vulnerabilities
```

If so, can you try checking if it still fails when turning it into a selection-only query, i.e.
```
FOR st IN PM_low_priority_assign
    FOR vuln IN PM_vulnerabilities
        FILTER vuln.lower_title LIKE st.cve_criteria
            RETURN {
                "special_team_title": st.team_assignment,
                "prime_contact": st.prime_contact,
                "comment": st.comment
            }
```
If this fails, then the LIMIT approach could be tried, by adding a LIMIT clause in front of RETURN and seeing if it fails for a specific document (or for a certain result set size).
Thanks.

**Comment by jsteemann** [comment]
@spirit-team : a few questions on this: 
* are the any special server options set in the config (i.e. `--query.*`)?
* are there are questions options set when running the query?
* is it possible that this is an out-of-memory error, i.e. is it possible that the server runs out of memory while this query is running? I am not saying this query will use a lot of memory, but it may be running on an already busy system. That's why I am asking. 
Thanks!

**Comment by jsteemann** [comment]
Without having a good way to reproduce the issue myself, I can only guess that the problem is somewhere in the UPDATE part. So here are a few ideas to try:

* when running the query from the arangosh, add `{ profile: 3 }` to the options, i.e. `db._query(queryString, null, { profile: 3 });`. That will write some data into the server log that may help find the part of the query that causes the issue
* when changing the UPDATE part of the query to just `UPDATE vuln WITH { } IN PM_vulnerabilities` the query should not update any documents, but still execute the UpdateExecutor. It could help to see if this succeeds or fails.
* wondering if it makes a difference which attributes are updated. For example, you could try updating one attribute first, to see if that works, e.g. first try `UPDATE vuln WITH {      "special_team_title": st.team_assignment } IN PM_vulnerabilities`. If that works, then check if updating `prime_contact` causes the issue or `comment`, or only all 3 together. Maybe there is one specific value that somehow cause trouble in the update.

If this doesn't produce any useful insights, a sensible next step from my point of view would be trying to reproduce the problem with an anonymized dataset or a small excerpt of the data, if possible.

**Comment by spirit-team** [comment]
@jsteemann Theres some good and not so great new here...
First of all I tried running the query  with { profile: 3 } as suggested and I didnt see anything in the logs that strike me as helpful but here is an excerpt anyway:
```
[36a75] 31.6665 Query::~Query queryString: this: 139849754116096
 a few seconds ago

[8bcee] slow query: 'FOR st IN PM_low_priority_assign2 FOR vuln IN PM_vulnerabilities FILTER REGEX_TEST(vuln.lower_title, st.cve_criteria, false) UPDATE vuln WITH { "special_team_title": st.team_assignment, "prime_contact": st.prime_contact, "co
 a few seconds ago

[1e717] [query#22534639122] execute type=EnumerateCollectionNode callStack= [ {specific: [ { skip: 0, softLimit: unlimited, hardLimit: unlimited, fullCount: false, skipCount: 0 } ]} ] this=139851758688256 id=3
 a few seconds ago

[1e717] [query#22534639122] execute type=CalculationNode callStack= [ {specific: [ { skip: 0, softLimit: unlimited, hardLimit: unlimited, fullCount: false, skipCount: 0 } ]} ] this=139852276721664 id=4
 a few seconds ago

[1e717] [query#22534639122] execute type=FilterNode callStack= [ {specific: [ { skip: 0, softLimit: unlimited, hardLimit: unlimited, fullCount: false, skipCount: 0 } ]} ] this=139852232092416 id=5
 a few seconds ago

[1e717] [query#22534639122] execute type=UpdateNode callStack= [ {specific: [ { skip: 0, softLimit: unlimited, hardLimit: unlimited, fullCount: false, skipCount: 0 } ]} ] this=139852276486144 id=7
 a few seconds ago

[60bbc] [query#22534639122] execute done type=UpdateNode this=139852276486144 id=7 state=HASMORE skipped=0 produced=1000 shadowRows=0
 a few seconds ago

[60bbc] [query#22534639122] execute done type=FilterNode this=139852232092416 id=5 state=HASMORE skipped=0 produced=1000 shadowRows=0
 a few seconds ago

[60bbc] [query#22534639122] execute done type=CalculationNode this=139852276721664 id=4 state=HASMORE skipped=0 produced=1000 shadowRows=0
 a few seconds ago

[60bbc] [query#22534639122] execute done type=EnumerateCollectionNode this=139851758688256 id=3 state=HASMORE skipped=0 produced=1000 shadowRows=0
 a few seconds ago

[1e717] [query#22534639122] execute type=EnumerateCollectionNode callStack= [ {specific: [ { skip: 0, softLimit: unlimited, hardLimit: unlimited, fullCount: false, skipCount: 0 } ]} ] this=139851758688256 id=3
 a few seconds ago

[1e717] [query#22534639122] execute type=CalculationNode callStack= [ {specific: [ { skip: 0, softLimit: unlimited, hardLimit: unlimited, fullCount: false, skipCount: 0 } ]} ] this=139852276721664 id=4
 a few seconds ago

[1e717] [query#22534639122] execute type=FilterNode callStack= [ {specific: [ { skip: 0, softLimit: unlimited, hardLimit: unlimited, fullCount: false, skipCount: 0 } ]} ] this=139852232092416 id=5
 a few seconds ago

[1e717] [query#22534639122] execute type=UpdateNode callStack= [ {specific: [ { skip: 0, softLimit: unlimited, hardLimit: unlimited, fullCount: false, skipCount: 0 } ]} ] this=139852276486144 id=7
 a few seconds ago

[60bbc] [query#22534639122] execute done type=UpdateNode this=139852276486144 id=7 state=HASMORE skipped=0 produced=1000 shadowRows=0
 a few seconds ago

[60bbc] [query#22534639122] execute done type=FilterNode this=139852232092416 id=5 state=HASMORE skipped=0 produced=1000 shadowRows=0
 a few seconds ago

[60bbc] [query#22534639122] execute done type=CalculationNode this=139852276721664 id=4 state=HASMORE skipped=0 produced=1000 shadowRows=0
 a few seconds ago

[60bbc] [query#22534639122] execute done type=EnumerateCollectionNode this=139851758688256 id=3 state=HASMORE skipped=0 produced=1000 shadowRows=0
 a few seconds ago

[1e717] [query#22534639122] execute type=EnumerateCollectionNode callStack= [ {specific: [ { skip: 0, softLimit: unlimited, hardLimit: unlimited, fullCount: false, skipCount: 0 } ]} ] this=139851758688256 id=3
 a few seconds ago

[1e717] [query#22534639122] execute type=CalculationNode callStack= [ {specific: [ { skip: 0, softLimit: unlimited, hardLimit: unlimited, fullCount: false, skipCount: 0 } ]} ] this=139852276721664 id=4
 a few seconds ago

[1e717] [query#22534639122] execute type=FilterNode callStack= [ {specific: [ { skip: 0, softLimit: unlimited, hardLimit: unlimited, fullCount: false, skipCount: 0 } ]} ] this=139852232092416 id=5
 a few seconds ago

[1e717] [query#22534639122] execute type=UpdateNode callStack= [ {specific: [ { skip: 0, softLimit: unlimited, hardLimit: unlimited, fullCount: false, skipCount: 0 } ]} ] this=139852276486144 id=7
 a few seconds ago

[60bbc] [query#22534639122] execute done type=UpdateNode this=139852276486144 id=7 state=HASMORE skipped=0 produced=1000 shadowRows=0
 a few seconds ago

[60bbc] [query#22534639122] execute done type=FilterNode this=139852232092416 id=5 state=HASMORE skipped=0 produced=1000 shadowRows=0
 a few seconds ago

[60bbc] [query#22534639122] execute done type=CalculationNode this=139852276721664 id=4 state=HASMORE skipped=0 produced=1000 shadowRows=0
 a few seconds ago

[60bbc] [query#22534639122] execute done type=EnumerateCollectionNode this=139851758688256 id=3 state=HASMORE skipped=0 produced=1000 shadowRows=0
 a few seconds ago

[1e717] [query#22534639122] execute type=EnumerateCollectionNode callStack= [ {specific: [ { skip: 0, softLimit: unlimited, hardLimit: unlimited, fullCount: false, skipCount: 0 } ]} ] this=139851758688256 id=3
 a few seconds ago

[1e717] [query#22534639122] execute type=CalculationNode callStack= [ {specific: [ { skip: 0, softLimit: unlimited, hardLimit: unlimited, fullCount: false, skipCount: 0 } ]} ] this=139852276721664 id=4
 a few seconds ago

[1e717] [query#22534639122] execute type=FilterNode callStack= [ {specific: [ { skip: 0, softLimit: unlimited, hardLimit: unlimited, fullCount: false, skipCount: 0 } ]} ] this=139852232092416 id=5
 a few seconds ago

[1e717] [query#22534639122] execute type=UpdateNode callStack= [ {specific: [ { skip: 0, softLimit: unlimited, hardLimit: unlimited, fullCount: false, skipCount: 0 } ]} ] this=139852276486144 id=7
 a few seconds ago
```
Its basicly very much like those for pages and pages..

Secondly, I tried replace the update part of my query with this UPDATE vuln WITH { } IN PM_vulnerabilities
And it worked, it completed in about 3.5 seconds.

Third, I tried updating only 1 field at a time and I did seperatly for all 3 fields, none of them completed successfully.

And finally the good news is that one of my colleague managed to get the query working by filtering early on the list of documents from PM_vulnerabilities.

```
LET vulnToUpdate = (
    FOR st IN PM_low_priority_assign2
        FOR vuln IN PM_vulnerabilities 
            FILTER CONTAINS(vuln.lower_title, LOWER(st.cve_criteria))
            RETURN DISTINCT vuln
)
FOR vuln IN vulnToUpdate
    FOR st IN PM_low_priority_assign2
    FILTER CONTAINS(vuln.lower_title, LOWER(st.cve_criteria))
        UPDATE vuln WITH { 
            "special_team_title": st.team_assignment, 
            "prime_contact": st.prime_contact, 
            "comment": st.comment 
        } IN PM_vulnerabilities
```
This is a nice work-around for us but doesnt resolve the issue that the other way of writing the query should work aswell.

I'll look into providing you with that anonymized data.

**Comment by spirit-team** [comment]
@jsteemann Finally found time to setup a 3.8.2 environment and test it. It is indeed FIXED in 3.8.2!!
Thanks alot.

Marking this issue as closed.

**Comment by jsteemann** [comment]
@spirit-team : nice, and good to read! Thanks for getting back!
